### PR TITLE
fix issue in bound service

### PR DESCRIPTION
**Pull request by ttt-77**

### CODE REVIEW

File: aidb/inference/bound_inference_service.py [hunk 1]
```diff
@@ -122,7 +122,13 @@ async def query_table():
...
             condition.append(getattr(self._cache_table.c, cache_col.name)
                              == getattr(self._tables[table_name]._table.c, col.name))
-      joined = sqlalchemy.join(joined, self._tables[table_name]._table, *condition)
+      if condition:
+        # Connected by 'AND' when the key is composite
+        join_condition = sqlalchemy.sql.and_(*condition)
+      else:
+        # Use CROSS JOIN in the absence of a specific condition.
+        join_condition = sqlalchemy.sql.true()
+      joined = sqlalchemy.join(joined, self._tables[table_name]._table, join_condition)
```
Comment (ttt-77) [inline aidb/inference/bound_inference_service.py:131]: fix the problem when the join key is a composite key

File: aidb/inference/bound_inference_service.py [hunk 2]
```diff
@@ -151,6 +157,7 @@ async def _insert_in_cache_table(self, inp_rows_df: pd.DataFrame, conn):
     # convert the pandas datatype to python native type
     # TODO: can we remove /resolve this?
     inp_rows_df = inp_rows_df.astype('object')
+    inp_rows_df = inp_rows_df.drop_duplicates()
```
Comment (ttt-77) [inline aidb/inference/bound_inference_service.py:160]: When sampling with replacement, a single element might be selected multiple times within a batch, potentially leading to insertion errors due to unique key constraints.

File: tests/inference_service_utils/inference_service_setup.py [hunk 3]
```diff
@@ -37,7 +37,7 @@ def register_inference_services(engine: Engine, data_dir: str, port=8000, batch_
       columns_to_input_keys=columns_to_input_keys,
       response_keys_to_columns=output_keys_to_columns,
       batch_supported=batch_supported,
-      preferred_batch_size=128,
+      preferred_batch_size=preferred_batch_size,
```
Comment (ttt-77) [inline tests/inference_service_utils/inference_service_setup.py:40]: support different batch size in different tests.

**Comment by ttt-77** [comment]
Could you please review this PR? @ddkang

File: aidb/inference/bound_inference_service.py [hunk 4]
```diff
@@ -122,7 +122,13 @@ async def query_table():
           if col.name == normal_name:
             condition.append(getattr(self._cache_table.c, cache_col.name)
                              == getattr(self._tables[table_name]._table.c, col.name))
-      joined = sqlalchemy.join(joined, self._tables[table_name]._table, *condition)
+      if condition:
```
Comment (ddkang) [inline aidb/inference/bound_inference_service.py:125]: probably more clear if you use len() != 0?

**Review by ddkang — COMMENTED** [review]
This PR looks okay, but I'm a bit worried about the caching. If AIDB runs like 1M inferences with 100k repeated, will we filter out the duplicates?

**Comment by ttt-77** [comment]
> This PR looks okay, but I'm a bit worried about the caching. If AIDB runs like 1M inferences with 100k repeated, will we filter out the duplicates?

When an element is already present in the cache, it will be ignored. When an element repeats multiple times within a single batch, it will be inserted only once.

**Comment by ddkang** [comment]
> > This PR looks okay, but I'm a bit worried about the caching. If AIDB runs like 1M inferences with 100k repeated, will we filter out the duplicates?
> 
> When an element is already present in the cache, it will be ignored. When an element repeats multiple times within a single batch, it will be inserted only once.

If a repeated element, will the _inference_ happen multiple times?

**Comment by ttt-77** [comment]
> > > This PR looks okay, but I'm a bit worried about the caching. If AIDB runs like 1M inferences with 100k repeated, will we filter out the duplicates?
> > 
> > 
> > When an element is already present in the cache, it will be ignored. When an element repeats multiple times within a single batch, it will be inserted only once.
> 
> If a repeated element, will the _inference_ happen multiple times?

If they are in the same batch, the _inference_ will happen multiple times. I'll think about how to modify it.

**Comment by ttt-77** [comment]
I now eliminate duplicates from the inputs instead of the outputs, ensuring that each inference is performed only once. Could you please review it? @ddkang

File: aidb/inference/bound_inference_service.py [hunk 5]
```diff
@@ -204,12 +210,15 @@ async def infer(self, inputs: pd.DataFrame):
     for idx, col in enumerate(self.binding.input_columns):
       inputs.rename(columns={inputs.columns[idx]: col}, inplace=True)
 
+    # Drop duplicate inputs
+    inputs_drop_duplicates = inputs.drop_duplicates()
```
Comment (ddkang) [inline aidb/inference/bound_inference_service.py:214]: How does this map back to the output? If the call requests records [1, 2, 1], will it return [1, 2, 1]?

Comment (ttt-77) [reply aidb/inference/bound_inference_service.py:214]: The current 'infer' function is designed not to return results directly. Instead, results are obtained through an external SQL query, ensuring that this modification does not affect the results. 
In another PR, I have revised the 'infer' function to directly return results. This is implemented by joining the input and output on common columns, thereby aligning the output size with the input size.

**Review by ddkang — APPROVED** [review]

**Pull request merged**